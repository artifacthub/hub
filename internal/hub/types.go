package hub

type userIDKey struct{}

// UserIDKey represents the key used for the userID value inside a context.
var UserIDKey = userIDKey{}

// ChartRepository represents a Helm chart repository.
type ChartRepository struct {
	ChartRepositoryID string `json:"chart_repository_id"`
	Name              string `json:"name"`
	DisplayName       string `json:"display_name"`
	URL               string `json:"url"`
	UserID            string `json:"user_id"`
}

// Link represents a url associated with a package.
type Link struct {
	Name string `json:"name"`
	URL  string `json:"url"`
}

// Maintainer represents a package's maintainer.
type Maintainer struct {
	MaintainerID string `json:"maintainer_id"`
	Name         string `json:"name"`
	Email        string `json:"email"`
}

// PackageKind represents the kind of a given package.
type PackageKind int64

const (
	// Chart represents a Helm chart.
	Chart PackageKind = 0

	// Falco represents a set of Falco rules.
	Falco PackageKind = 1

	// OPA represents a set of OPA policies.
	OPA PackageKind = 2
)

// Package represents a Kubernetes package.
type Package struct {
	PackageID         string                 `json:"package_id"`
	Kind              PackageKind            `json:"kind"`
	Name              string                 `json:"name"`
	DisplayName       string                 `json:"display_name"`
	Description       string                 `json:"description"`
	HomeURL           string                 `json:"home_url"`
	LogoURL           string                 `json:"logo_url"`
	LogoImageID       string                 `json:"logo_image_id"`
	Keywords          []string               `json:"keywords"`
	Deprecated        bool                   `json:"deprecated"`
	Readme            string                 `json:"readme"`
	Links             []*Link                `json:"links"`
	Version           string                 `json:"version"`
	AvailableVersions []string               `json:"available_versions"`
	AppVersion        string                 `json:"app_version"`
	Digest            string                 `json:"digest"`
	Data              map[string]interface{} `json:"data"`
	Maintainers       []*Maintainer          `json:"maintainers"`
	ChartRepository   *ChartRepository       `json:"chart_repository"`
}

// SearchPackageInput represents the query input when searching for packages.
type SearchPackageInput struct {
	Limit             int           `json:"limit,omitempty"`
	Offset            int           `json:"offset,omitempty"`
	Facets            bool          `json:"facets"`
	Text              string        `json:"text"`
	PackageKinds      []PackageKind `json:"package_kinds,omitempty"`
	ChartRepositories []string      `json:"chart_repositories,omitempty"`
	Deprecated        bool          `json:"deprecated"`
}

// GetPackageInput represents the input used to get a specific package.
type GetPackageInput struct {
	ChartRepositoryName string `json:"chart_repository_name"`
	PackageName         string `json:"package_name"`
	Version             string `json:"version"`
}

// User represents a Hub user.
type User struct {
	UserID        string `json:"user_id"`
	Alias         string `json:"alias"`
	FirstName     string `json:"first_name"`
	LastName      string `json:"last_name"`
	Email         string `json:"email"`
	EmailVerified bool   `json:"email_verified"`
	Password      string `json:"password"`
}

// CheckCredentialsOutput represents the output returned by the
// CheckCredentials method.
type CheckCredentialsOutput struct {
	Valid  bool   `json:"valid"`
	UserID string `json:"user_id"`
}

// Session represents some information about a user session.
type Session struct {
	SessionID string `json:"session_id"`
	UserID    string `json:"user_id"`
	IP        string `json:"ip"`
	UserAgent string `json:"user_agent"`
}

// CheckSessionOutput represents the output returned by the CheckSession method.
type CheckSessionOutput struct {
	Valid  bool   `json:"valid"`
	UserID string `json:"user_id"`
}

// Organization represents an entity with one or more users associated that can
// own packages and other entities like chart repositories.
type Organization struct {
	OrganizationID string `json:"organization_id"`
	Name           string `json:"name"`
	DisplayName    string `json:"display_name"`
	Description    string `json:"description"`
	HomeURL        string `json:"home_url"`
	LogoURL        string `json:"logo_url"`
	LogoImageID    string `json:"logo_image_id"`
}
