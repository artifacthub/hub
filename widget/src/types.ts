export interface PackageSummary {
  packageId: string;
  name: string;
  displayName?: string;
  normalizedName: string;
  description: string;
  logoImageId?: string;
  repository: Repository;
  version: string;
  stars?: number | null;
  ts: number;
  official?: boolean;
}

export interface Repository {
  kind: RepositoryKind;
  name: string;
  verifiedPublisher?: boolean;
  official?: boolean;
  organizationName?: string;
  organizationDisplayName?: string;
  userAlias?: string;
}

export enum RepositoryKind {
  Helm = 0,
  Falco,
  OPA,
  OLM,
  TBAction,
  Krew,
  HelmPlugin,
  TektonTask,
  KedaScaler,
  CoreDNS,
  Keptn,
  TektonPipeline,
  Container,
  Kubewarden,
}

export interface SearchResults {
  packages: PackageSummary[] | null;
  facets: null;
}
