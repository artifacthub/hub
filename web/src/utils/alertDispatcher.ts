import isUndefined from 'lodash/isUndefined';

import { Alert } from '../types';

const DEFUALT_TIME = 5 * 1000; //5s

export interface AlertUpdatesHandler {
  updateAlertWrapper(alert: Alert | null): void;
}

export class AlertDispatcher {
  private updatesHandler?: AlertUpdatesHandler;
  private visibleAlert: Alert | null = null;
  private timeout?: NodeJS.Timeout;

  public subscribe(updatesHandler: AlertUpdatesHandler) {
    this.updatesHandler = updatesHandler;
  }

  public postAlert(alert: Alert | null) {
    this.visibleAlert = alert;
    this.dismissAlert();
    if (this.updatesHandler) {
      this.updatesHandler.updateAlertWrapper(this.visibleAlert);
    }
  }

  private dismissAlert() {
    this.clearTimeout();
    if (this.visibleAlert && (isUndefined(this.visibleAlert.autoClose) || this.visibleAlert.autoClose)) {
      this.timeout = setTimeout(() => {
        this.postAlert(null);
      }, this.visibleAlert.dismissOn || DEFUALT_TIME);
    }
  }

  private clearTimeout() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  }
}

const alertDispatcher = new AlertDispatcher();
export default alertDispatcher;
