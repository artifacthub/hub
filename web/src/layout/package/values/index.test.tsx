import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mocked } from 'jest-mock';
import { BrowserRouter as Router } from 'react-router-dom';

import API from '../../../api';
import { ErrorKind } from '../../../types';
import alertDispatcher from '../../../utils/alertDispatcher';
import Values from './';
jest.mock('../../../api');
jest.mock('../../../utils/alertDispatcher');

const mockUseNavigate = jest.fn();
const scrollToMock = jest.fn();
const itemScrollMock = jest.fn();

Object.defineProperty(HTMLElement.prototype, 'scroll', { configurable: true, value: itemScrollMock });
window.HTMLElement.prototype.scrollTo = scrollToMock;

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as {}),
  useNavigate: () => mockUseNavigate,
}));

jest.mock('react-syntax-highlighter', () => () => <div id="line_59">port:</div>);

const defaultProps = {
  packageId: 'id',
  version: '0.1.0',
  visibleValues: false,
  normalizedName: 'pkg',
  sortedVersions: [
    { version: '1.1.1', ts: 1, containsSecurityUpdates: false, prerelease: false },
    { version: '1.1.0', ts: 1, containsSecurityUpdates: false, prerelease: false },
    { version: '1.0.1', ts: 1, containsSecurityUpdates: false, prerelease: false },
    { version: '1.0.0', ts: 1, containsSecurityUpdates: false, prerelease: false },
  ],
};

const YAMLSample = `nameOverride: ""
fullnameOverride: ""
imagePullSecrets: []
imageTag: ""
dynamicResourceNamePrefixEnabled: false
pullPolicy: IfNotPresent
restrictedHTTPClient: false

log:
  level: info
  pretty: false

db:
  host: ""
  port: "5432"
  database: hub
  user: postgres
  password: postgres

email:
  fromName: ""
  from: ""
  replyTo: ""
  smtp:
    auth: plain
    host: ""
    port: 587
    username: ""
    password: ""

creds:
  dockerUsername: ""
  dockerPassword: ""
  githubToken: ""

images:
  store: pg

events:
  scanningErrors: false
  trackingErrors: false

dbMigrator:
  job:
    image:
      repository: artifacthub/db-migrator
  loadSampleData: true
  configDir: "/home/db-migrator/.cfg"

hub:
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: nginx
    rules: []
    tls: []
  service:
    type: NodePort
    port: 80
  deploy:
    readinessGates: []
    replicaCount: 1
    image:
      repository: artifacthub/hub
    resources: {}
  server:
    allowPrivateRepositories: false
    allowUserSignUp: true
    cacheDir: ""
    configDir: "/home/hub/.cfg"
    baseURL: ""
    shutdownTimeout: 10s
    motd: ""
    motdSeverity: info
    basicAuth:
      enabled: false
      username: hub
      password: changeme
    cookie:
      hashKey: default-unsafe-key
      secure: false
    csrf:
      authKey: default-unsafe-key
      secure: false
    oauth:
      github:
        enabled: false
        clientID: ""
        clientSecret: ""
        redirectURL: ""
        scopes:
          - read:user
          - user:email
      google:
        enabled: false
        clientID: ""
        clientSecret: ""
        redirectURL: ""
        scopes:
          - https://www.googleapis.com/auth/userinfo.email
          - https://www.googleapis.com/auth/userinfo.profile
      oidc:
        enabled: false
        issuerURL: ""
        clientID: ""
        clientSecret: ""
        redirectURL: ""
        scopes:
          - openid
          - profile
          - email
        skipEmailVerifiedCheck: false
    xffIndex: 0
  analytics:
    gaTrackingID: ""
  theme:
    colors:
      primary: "#417598"
      secondary: "#2D4857"
    images:
      appleTouchIcon192: "/static/media/logo192_v2.png"
      appleTouchIcon512: "/static/media/logo512_v2.png"
      openGraphImage: "/static/media/artifactHub_v2.png"
      shortcutIcon: "/static/media/logo_v2.png"
      websiteLogo: "/static/media/logo/artifacthub-brand-white.svg"
    sampleQueries: []
    siteName: "Artifact hub"

scanner:
  cronjob:
    image:
      repository: artifacthub/scanner
    resources: {}
  concurrency: 10
  trivyURL: ""
  cacheDir: ""
  configDir: "/home/scanner/.cfg"

tracker:
  cronjob:
    image:
      repository: artifacthub/tracker
    resources: {}
  cacheDir: ""
  configDir: "/home/tracker/.cfg"
  concurrency: 10
  repositoriesNames: []
  repositoriesKinds: []
  bypassDigestCheck: false

trivy:
  deploy:
    image: aquasec/trivy:0.20.2
    resources: {}
  persistence:
    enabled: false
    size: 10Gi

# Values for postgresql chart dependency
postgresql:
  enabled: true
  image:
    repository: postgres
    tag: 12
  persistence:
    mountPath: /data
  postgresqlUsername: postgres
  postgresqlPassword: postgres
  postgresqlDatabase: hub
  postgresqlDataDir: /data/pgdata
`;

describe('Values', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    mocked(API).getChartValues.mockResolvedValue(YAMLSample);

    const { asFragment } = render(
      <Router>
        <Values {...defaultProps} visibleValues />
      </Router>
    );

    await waitFor(() => {
      expect(API.getChartValues).toHaveBeenCalledTimes(1);
    });

    await waitFor(() => expect(screen.getByRole('dialog')).toBeInTheDocument());
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', async () => {
      mocked(API).getChartValues.mockResolvedValue(YAMLSample);

      render(
        <Router>
          <Values {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: /Open default values modal/ });
      expect(btn).toBeInTheDocument();
      await userEvent.click(btn);

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(1);
        expect(API.getChartValues).toHaveBeenCalledWith(defaultProps.packageId, defaultProps.version);
      });

      await waitFor(() => expect(screen.getByRole('dialog')).toBeInTheDocument());
    });

    it('opens modal', async () => {
      mocked(API).getChartValues.mockResolvedValue(YAMLSample);

      render(
        <Router>
          <Values {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: /Open default values modal/ });
      await userEvent.click(btn);

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(1);
        expect(mockUseNavigate).toHaveBeenCalledTimes(1);
        expect(mockUseNavigate).toHaveBeenCalledWith(
          {
            search: '?modal=values',
          },
          { replace: true, state: null }
        );
      });

      await waitFor(() => expect(screen.getByRole('dialog')).toBeInTheDocument());
      expect(screen.getAllByText('Default values')).toHaveLength(2);
    });

    it('closes modal', async () => {
      mocked(API).getChartValues.mockResolvedValue(YAMLSample);

      render(
        <Router>
          <Values {...defaultProps} visibleValues />
        </Router>
      );

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(1);
      });

      expect(await screen.findByRole('dialog')).toBeInTheDocument();

      const close = screen.getByRole('button', { name: 'Close modal' });
      await userEvent.click(close);

      expect(mockUseNavigate).toHaveBeenCalledTimes(1);
      expect(mockUseNavigate).toHaveBeenCalledWith('', { replace: true, state: null });

      await waitFor(() => expect(screen.queryByRole('dialog')).toBeNull());
    });

    it('opens modal with empty values', async () => {
      mocked(API).getChartValues.mockResolvedValue('');

      render(
        <Router>
          <Values {...defaultProps} visibleValues />
        </Router>
      );

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(1);
      });

      expect(await screen.findByRole('dialog')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Copy to clipboard' })).toBeDisabled();
      expect(screen.getByRole('button', { name: 'Download' })).toBeDisabled();
    });

    it('calls again to getChartValues when version is different', async () => {
      mocked(API).getChartValues.mockResolvedValue(YAMLSample);

      const { rerender } = render(
        <Router>
          <Values {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: /Open default values modal/ });
      await userEvent.click(btn);

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(1);
        expect(API.getChartValues).toHaveBeenCalledWith(defaultProps.packageId, defaultProps.version);
      });

      expect(await screen.findByRole('dialog')).toBeInTheDocument();

      const close = screen.getByText('Close');
      await userEvent.click(close);

      rerender(
        <Router>
          <Values {...defaultProps} version="1.0.0" />
        </Router>
      );

      await userEvent.click(btn);

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(2);
        expect(API.getChartValues).toHaveBeenCalledWith(defaultProps.packageId, '1.0.0');
      });

      expect(await screen.findByRole('dialog')).toBeInTheDocument();
    });

    it('calls again to getChartValues when packageId is different', async () => {
      mocked(API).getChartValues.mockResolvedValue(YAMLSample);

      const { rerender } = render(
        <Router>
          <Values {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: /Open default values modal/ });
      await userEvent.click(btn);

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(1);
        expect(API.getChartValues).toHaveBeenCalledWith(defaultProps.packageId, defaultProps.version);
      });

      expect(await screen.findByRole('dialog')).toBeInTheDocument();

      const close = screen.getByText('Close');
      await userEvent.click(close);

      rerender(
        <Router>
          <Values {...defaultProps} packageId="id2" />
        </Router>
      );

      await userEvent.click(btn);

      await waitFor(() => {
        expect(API.getChartValues).toHaveBeenCalledTimes(2);
        expect(API.getChartValues).toHaveBeenCalledWith('id2', defaultProps.version);
      });

      expect(await screen.findByRole('dialog')).toBeInTheDocument();
    });

    it('closes modal when a new pkg is open', async () => {
      mocked(API).getChartValues.mockResolvedValue(YAMLSample);

      const { rerender } = render(
        <Router>
          <Values {...defaultProps} />
        </Router>
      );

      const btn = screen.getByRole('button', { name: /Open default values modal/ });
      await userEvent.click(btn);

      expect(await screen.findByRole('dialog')).toBeInTheDocument();
      expect(API.getChartValues).toHaveBeenCalledTimes(1);

      rerender(
        <Router>
          <Values {...defaultProps} packageId="id2" />
        </Router>
      );

      await waitFor(() => expect(screen.queryByRole('dialog')).toBeNull());
    });

    describe('search', () => {
      it('renders search bar', async () => {
        mocked(API).getChartValues.mockResolvedValue(YAMLSample);

        render(
          <Router>
            <Values {...defaultProps} visibleValues />
          </Router>
        );

        await waitFor(() => {
          expect(API.getChartValues).toHaveBeenCalledTimes(1);
        });

        expect(await screen.findByRole('dialog')).toBeInTheDocument();

        await userEvent.type(screen.getByRole('textbox'), 'name');

        const opts = await screen.findAllByTestId('typeaheadDropdownBtn');
        expect(opts).toHaveLength(10);
        expect(opts[0]).toHaveTextContent('nameOverride');
        expect(opts[1]).toHaveTextContent('fullnameOverride');
        expect(opts[2]).toHaveTextContent('dynamicResourceNamePrefixEnabled');
        expect(opts[3]).toHaveTextContent('email.fromName');
        expect(opts[4]).toHaveTextContent('email.smtp.username');
        expect(opts[5]).toHaveTextContent('creds.dockerUsername');
        expect(opts[6]).toHaveTextContent('hub.server.basicAuth.username');
        expect(opts[7]).toHaveTextContent('hub.theme.siteName');
        expect(opts[8]).toHaveTextContent('tracker.repositoriesNames');
        expect(opts[9]).toHaveTextContent('postgresql.postgresqlUsername');
      });

      it('clicks option after searching', async () => {
        mocked(API).getChartValues.mockResolvedValue(YAMLSample);

        render(
          <Router>
            <Values {...defaultProps} visibleValues />
          </Router>
        );

        await waitFor(() => {
          expect(API.getChartValues).toHaveBeenCalledTimes(1);
        });

        expect(await screen.findByRole('dialog')).toBeInTheDocument();

        await userEvent.type(screen.getByRole('textbox'), 'name');

        const opts = screen.getAllByTestId('typeaheadDropdownBtn');

        await userEvent.click(opts[8]);

        await waitFor(() => {
          expect(mockUseNavigate).toHaveBeenCalledTimes(1);
          expect(mockUseNavigate).toHaveBeenCalledWith(
            { search: '?modal=values&path=tracker.repositoriesNames' },
            {
              replace: true,
              state: null,
            }
          );
        });
      });

      it('goes to correct position when querystring path is defined', async () => {
        mocked(API).getChartValues.mockResolvedValue(YAMLSample);

        render(
          <Router>
            <Values {...defaultProps} visibleValues visibleValuesPath="hub.service.port" />
          </Router>
        );

        await waitFor(() => {
          expect(API.getChartValues).toHaveBeenCalledTimes(1);
        });

        expect(await screen.findByRole('dialog')).toBeInTheDocument();

        await waitFor(() => {
          expect(scrollToMock).toHaveBeenCalledTimes(1);
        });
      });
    });

    describe('when fails', () => {
      it('on NotFound', async () => {
        mocked(API).getChartValues.mockRejectedValue({
          kind: ErrorKind.NotFound,
        });

        render(
          <Router>
            <Values {...defaultProps} />
          </Router>
        );

        const btn = screen.getByRole('button', { name: /Open default values modal/ });
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.getChartValues).toHaveBeenCalledTimes(1);
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'danger',
            message:
              'We could not find the default values for this chart version. Please check that the chart tgz package still exists in the source repository as it might not be available anymore.',
            dismissOn: 10000,
          });
        });
      });

      it('default error', async () => {
        mocked(API).getChartValues.mockRejectedValue({ kind: ErrorKind.Other });

        render(
          <Router>
            <Values {...defaultProps} />
          </Router>
        );

        const btn = screen.getByRole('button', { name: /Open default values modal/ });
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.getChartValues).toHaveBeenCalledTimes(1);
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'danger',
            message: 'An error occurred getting chart default values, please try again later.',
          });
        });

        await waitFor(() => {
          expect(mockUseNavigate).toHaveBeenCalledTimes(2);
          expect(mockUseNavigate).toHaveBeenLastCalledWith('', { replace: true, state: null });
        });
      });
    });
  });
});
