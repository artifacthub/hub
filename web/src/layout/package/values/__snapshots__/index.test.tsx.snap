// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Values creates snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="Open default values modal"
    class="btn btn-outline-secondary btn-sm w-100"
  >
    <div
      class="d-flex flex-row align-items-center justify-content-center text-uppercase"
    >
      <svg
        class="me-2"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="1"
          width="9"
          x="4"
          y="9"
        />
        <rect
          height="1"
          width="7"
          x="4"
          y="12"
        />
        <rect
          height="1"
          width="10"
          x="4"
          y="6"
        />
        <rect
          height="1"
          width="11"
          x="1"
          y="3"
        />
        <rect
          height="9"
          width="1"
          x="4"
          y="4"
        />
      </svg>
      <span
        class="fw-bold"
      >
        Default values
      </span>
    </div>
  </button>
  <div>
    <div
      class="modal-backdrop activeBackdrop"
      data-testid="modalBackdrop"
    />
    <div
      aria-modal="true"
      class="modal modal active d-block"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable modalDialog"
      >
        <div
          class="modal-content border border-3 mx-auto content h-100"
        >
          <div
            class="modal-header d-flex flex-row align-items-center header undefined"
          >
            <div
              class="d-flex flex-row align-items-center flex-grow-1"
            >
              <div
                class="h3 m-2 flex-grow-1 title"
              >
                Default values
              </div>
              <div
                class="mx-4"
              >
                <div
                  class="btn-group"
                  role="group"
                >
                  <div
                    class="position-relative undefined"
                  >
                    <div
                      data-testid="elementWithTooltip"
                    >
                      <button
                        aria-disabled="false"
                        aria-label="Open diff template view"
                        class="btn btn-outline-primary btn-sm dropdown-toggle"
                        type="button"
                      >
                        <span
                          class="pe-2"
                        >
                          Compare to version ...
                        </span>
                      </button>
                    </div>
                  </div>
                  <div
                    class="dropdown-menu dropdown-menu-end text-nowrap overflow-hidden versionsDropdown"
                    role="complementary"
                  >
                    <div
                      class="overflow-scroll h-100"
                    >
                      <button
                        class="dropdown-item btnItem"
                      >
                        <div
                          class="d-flex flex-row align-items-center"
                        >
                          <div
                            class="itemIcon"
                          />
                          <div
                            class="ms-2 text-truncate"
                          >
                            1.1.1
                          </div>
                        </div>
                      </button>
                      <button
                        class="dropdown-item btnItem"
                      >
                        <div
                          class="d-flex flex-row align-items-center"
                        >
                          <div
                            class="itemIcon"
                          />
                          <div
                            class="ms-2 text-truncate"
                          >
                            1.1.0
                          </div>
                        </div>
                      </button>
                      <button
                        class="dropdown-item btnItem"
                      >
                        <div
                          class="d-flex flex-row align-items-center"
                        >
                          <div
                            class="itemIcon"
                          />
                          <div
                            class="ms-2 text-truncate"
                          >
                            1.0.1
                          </div>
                        </div>
                      </button>
                      <button
                        class="dropdown-item btnItem"
                      >
                        <div
                          class="d-flex flex-row align-items-center"
                        >
                          <div
                            class="itemIcon"
                          />
                          <div
                            class="ms-2 text-truncate"
                          >
                            1.0.0
                          </div>
                        </div>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          >
            <div
              class="mw-100 h-100 d-flex flex-column overflow-hidden"
            >
              <div
                class="d-flex flex-row"
              >
                <div
                  class="position-relative wrapper w-50"
                >
                  <div
                    class="mb-3 input-group-sm inputWrapper undefined"
                  >
                    <input
                      class="flex-grow-1 form-control input ps-3 pe-4"
                      name="inputTypeahead_path"
                      placeholder="Search path"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <svg
                      class="text-muted position-absolute searchIcon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="position-relative flex-grow-1 content"
              >
                <div
                  class="position-absolute d-flex flex-row wrapper undefined"
                >
                  <div
                    class="position-relative me-2"
                  >
                    <button
                      aria-label="Copy to clipboard"
                      class="btn btn-sm btn-primary btn"
                      type="button"
                    >
                      <div
                        aria-hidden="true"
                        class="d-flex flex-row align-items-center"
                      >
                        <svg
                          fill="none"
                          height="1em"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <rect
                            height="13"
                            rx="2"
                            ry="2"
                            width="13"
                            x="9"
                            y="9"
                          />
                          <path
                            d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
                          />
                        </svg>
                      </div>
                    </button>
                  </div>
                  <button
                    aria-label="Download"
                    class="btn btn-sm btn-primary fs-5 btn"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M18.948 11.112C18.511 7.67 15.563 5 12.004 5c-2.756 0-5.15 1.611-6.243 4.15-2.148.642-3.757 2.67-3.757 4.85 0 2.757 2.243 5 5 5h1v-2h-1c-1.654 0-3-1.346-3-3 0-1.404 1.199-2.757 2.673-3.016l.581-.102.192-.558C8.153 8.273 9.898 7 12.004 7c2.757 0 5 2.243 5 5v1h1c1.103 0 2 .897 2 2s-.897 2-2 2h-2v2h2c2.206 0 4-1.794 4-4a4.008 4.008 0 0 0-3.056-3.888z"
                      />
                      <path
                        d="M13.004 14v-4h-2v4h-3l4 5 4-5z"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  class="overflow-auto h-100 position-relative border border-1 codeWrapper"
                >
                  <div
                    id="line_59"
                  >
                    port:
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="overflow-hidden alertWrapper"
                data-testid="alertWrapper"
              />
            </div>
          </div>
          <div
            class="modal-footer p-3 modalFooter"
          >
            <button
              aria-label="Close modal"
              class="btn btn-sm btn-outline-secondary text-uppercase"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
